c++/src/kj/compat/http.c++

```diff
@@ -3490,7 +3549,260 @@ WebSocketPipe newWebSocketPipe() {
 // =======================================================================================
 
 namespace {
+  class WebSocketCompressionUtil {
+    // A utility class with string parsing methods for processing `Sec-WebSocket-Extensions`.
+  public:
+    struct CompressionParameters {
+      // These are the parameters for `Sec-WebSocket-Extensions` permessage-deflate extension.
+      // Since we cannot distinguish the client/server in `upgradeToWebSocket`, we use the prefixes
+      // `inbound` and `outbound` instead.
+      bool outbound_no_context_takeover = false;
+      bool inbound_no_context_takeover = false;
+      kj::Maybe<size_t> outbound_max_window_bits = nullptr;
+      kj::Maybe<size_t> inbound_max_window_bits = nullptr;
+    };
+    static kj::String generateExtensionResponse(const CompressionParameters& parameters);
+    static kj::Maybe<CompressionParameters> validateRequestParameters(
+        kj::ArrayPtr<kj::String> parameters);
+    // Server functions
+
+    static kj::Maybe<CompressionParameters> retainClientCompressorConfig(const kj::StringPtr offer);
+    static kj::Maybe<CompressionParameters> validateResponseParameters(
+        kj::ArrayPtr<kj::String> parameters);
+    // Client functions
+  };
+
+  kj::Maybe<WebSocketCompressionUtil::CompressionParameters>
+      WebSocketCompressionUtil::validateRequestParameters(kj::ArrayPtr<kj::String> parameters) {
+    if (parameters.size() > 5) {
+      // `permessage-deflate` has 4 optional parameters. If we've seen all 4 parameters, then any
+      // additional parameters are either repeats, or invalid, so there's no need to process them.
+      // `parameters` starts with `permessage-deflate`, so we should have 5 elements at most.
+      return nullptr;
+    }
+
+    kj::Maybe<WebSocketCompressionUtil::CompressionParameters> compressionParameters =
+        WebSocketCompressionUtil::CompressionParameters {};
+    auto& config = KJ_REQUIRE_NONNULL(compressionParameters);
+
+    for (const auto &i : kj::indices(parameters)) {
+      if (i == 0) {
+        // Skip first element, which is `permessage-deflate`.
+        continue;
+      }
+      if (parameters[i].startsWith("server_max_window_bits=")) {
+        // TODO (now): RFC seems to imply that if the param has a value, it will not have
```

Comment:
Not completely certain about this bit. RFC-6455 [section 9.1](https://datatracker.ietf.org/doc/html/rfc6455#section-9.1) says extension parameters have the following abnf:
```
extension-param = token [ "=" (token | quoted-string) ]
```
So it's a bit unclear to me (for now) whether `token *LWS "=" *LWS (token | quoted-string)` is valid. Ex. can we have `client_max_window_bits       =   10`? It isn't necessarily a problem since if we fail to determine the value, we'll default to the max window as the server, and `min(offer, max_window)` as the client.

Reply:
Fixup https://github.com/capnproto/capnproto/pull/1475/commits/7da637bd2b26d3205108b11851cf3532b5425bc5